File: scala/lemkit-model/src/main/scala/io/people8/classify/PredictApp.scala
package io.people8.classify

import scala.util.control.Breaks._

import data._

object PredictApp extends App {
  var modelFormat = "binary"
  var predictFile: Option[String] = None
  var readModel: Option[String] = None
  var showCorrect = false
  var showAccuracy = false
  type FeatObs = FeatureObservation[String]
  type Examp = Example[String, Seq[FeatObs]]

  var i = 0
  val len = args.size

  def getarg() = {
    i += 1
    require(i < len, s"Argument for ${args(i - 1)} required")
    args(i)
  }
  def getChoice(choices: Seq[String]) = {
    val value = getarg()
    require(choices.contains(value),
      "Argument for %s must be %s" format (args(i - 1),
        choices.map(c => s"'$c'").mkString(" or ")))
    value
  }

  breakable {
    while (i < len) {
      args(i) match {
        case "--model-format" | "--mf" | "-f" => {
          modelFormat = getChoice(Seq("json", "binary"))
        }
        case "--predict" | "-p" => {
          predictFile = Some(getarg())
        }
        case "--model" | "-m" => {
          readModel = Some(getarg())
        }
        case "--show-accuracy" | "-a" => {
          showAccuracy = true
        }
        case "--show-correct" | "-c" => {
          showCorrect = true
        }
        case "--help" | "-h" => {
          println("""Usage:

    --model-format | --mf | -f  Model format (json or binary, default binary)
    --predict | -p              File containing data instances to predict
    --model | -m                Trained model file
    --show-accuracy | -a        Output accuracy at end
    --show-correct | -c         Output column indicating correct or wrong""")
        }
        case _ => {
          // If we took positional arguments, we would eliminate the
          // error here and process them down below
          throw new IllegalArgumentException("Unrecognized arguments: %s" format
            args.slice(i, len).mkString(" "))
        }
      }
      i += 1
    }
  }

  object IdentityFeaturizer extends Featurizer[Seq[FeatObs], String] {
    def apply(input: Seq[FeatObs]) = input
  }

  val classifier =
    readModel match {
      case Some(file) => {
        if (modelFormat == "json")
          LinearClassifier.readJSONModel(file, IdentityFeaturizer)
        else
          LinearClassifier.readBinaryModel(file, IdentityFeaturizer)
      }
      case None =>
        throw new IllegalArgumentException("Must specify --model (or -m)")
    }

  predictFile match {
    case Some(file) => {
      val predictData = ClassifierSource.readDataFile(file).toSeq
      val predictions = predictData.map(i => classifier(i.features))
      val numinsts = predictData.size
      var numcorrect = 0
      for (
        ((prediction, inst), index) <- (predictions zip predictData).zipWithIndex
      ) {
        val correct = inst.label
        val isCorrect = correct == prediction
        println("%s%s %s %s" format (
          index + 1,
          if (!showCorrect) "" else if (isCorrect) " CORRECT" else " WRONG",
          correct, prediction))
        if (isCorrect)
          numcorrect += 1
      }
      if (showAccuracy)
        println("Accuracy: %.2f%%" format (numcorrect.toDouble * 100 / numinsts))
    }
    case None =>
      throw new IllegalArgumentException("Must specify --test (or -t)")
  }
}

File: go/predict/PredictApp.go
package main

import (
	"flag"
	"fmt"
	"github.com/peoplepattern/commons-go/classify"
	"os"
)

func init() {
	flag.CommandLine = flag.NewFlagSet(os.Args[0], flag.ExitOnError)
}

var modelFormat = flag.String("model-format", "binary", "Model format (json or binary, default binary)")
var modelFile = flag.String("model", "", "Trained model file")
var predictFile = flag.String("predict", "", "File containing data instances to predict")
var showAccuracy = flag.Bool("show-accuracy", false, "Output accuracy at end")
var showCorrect = flag.Bool("show-correct", false, "Output column indicating correct or wrong")

func init() {
	flag.StringVar(modelFormat, "f", "binary", "Model format (json or binary, default binary)")
	flag.StringVar(modelFile, "m", "", "Trained model file")
	flag.StringVar(predictFile, "p", "", "File containing data instances to predict")
	flag.BoolVar(showAccuracy, "a", false, "Output accuracy at end")
	flag.BoolVar(showCorrect, "c", false, "Output column indicating correct or wrong")
}

func main() {
	flag.Parse()
	if *modelFormat != "json" && *modelFormat != "binary" {
		fmt.Printf("Invalid value for -model-format: %s\n", *modelFormat)
		os.Exit(1)
	}
	if *modelFile == "" {
		fmt.Println("Must specify -model (or -m)")
		os.Exit(1)
	}
	if *predictFile == "" {
		fmt.Println("Must specify -predict (or -p)")
		os.Exit(1)
	}

	var classifier *classify.LinearClassifier
	var err error
	if *modelFormat == "json" {
		classifier, err = classify.ReadJSONModel(*modelFile)
	} else {
		classifier, err = classify.ReadBinaryModel(*modelFile)
	}
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	predictData, err := classify.ReadDataFile(*predictFile)
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	numinsts := len(predictData)
	predictions := make([]string, numinsts)
	for i := 0; i < numinsts; i++ {
		predictions[i], _ = classifier.Predict(predictData[i].Features)
	}
	numcorrect := 0
	for i := 0; i < numinsts; i++ {
		correct := predictData[i].Label
		prediction := predictions[i]
		isCorrect := correct == prediction
		correctString := ""
		if *showCorrect {
			if isCorrect {
				correctString = " CORRECT"
			} else {
				correctString = " WRONG"
			}
		}
		fmt.Printf("%d%s %s %s\n", i+1, correctString, correct, prediction)
		if isCorrect {
			numcorrect++
		}
	}
	if *showAccuracy {
		fmt.Printf("Accuracy: %.2f%%\n", float64(numcorrect)*100.0/float64(numinsts))
	}
}

// For Vim, so we get 4-space tabs
// vim: set ts=4 sw=4 noet:
